SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-4028/SpatialAuthorityTest.cpp

```diff
@@ -0,0 +1,358 @@
+			[this](float DeltaTime)
+			{
+				if (DynamicReplicatedActor != nullptr
+					&& DynamicReplicatedActor->AuthorityOnBeginPlay == 1
+					&& DynamicReplicatedActor->AuthorityOnTick == 1)
+				{
+					FinishStep();
+				}
+			}, 5.0f);
+
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Destroy"), FWorkerDefinition::Server(1), nullptr,
+			[this]()
+			{
+				DynamicReplicatedActor->Destroy();
+				CrossServerSetDynamicReplicatedActor(nullptr);
+				FinishStep();
+			});
+	}
+
+	// Replicated Dynamic Actor Spawned On Different Server. Server 1 should have Authority on BeginPlay, Server 2 on Tick
+	{
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Different Server - Spawn"), FWorkerDefinition::Server(1), nullptr,
+				[this]() {
+					ASpatialAuthorityTestReplicatedActor* Actor = GetWorld()->SpawnActor<ASpatialAuthorityTestReplicatedActor>(
+						Server2Position, FRotator::ZeroRotator);
+					CrossServerSetDynamicReplicatedActor(Actor);
+					FinishStep();
+				});
+
+		AddStep(
+			TEXT("Replicated Dynamic Actor Spawned On Different Server - Verify Server 1 Has Authority on BeginPlay and Server 2 on Tick"), FWorkerDefinition::AllWorkers, nullptr,
+			nullptr,
+			[this](float DeltaTime) {
+				// Allow it to continue working in Native / Single worker setups.
+				int ExpectedAuthorityServer = GetNumberOfServerWorkers() > 1 ? 2 : 1;				
+				if (DynamicReplicatedActor != nullptr
+					&& DynamicReplicatedActor->AuthorityOnBeginPlay == 1
+					&& DynamicReplicatedActor->AuthorityOnTick == ExpectedAuthorityServer)
+				{
+					FinishStep();
+				}
+			},
+			5.0f);
+
+		// Now it's Server 2 destroying, since it has Authority over it.
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Different Server - Destroy"), FWorkerDefinition::Server(2), nullptr,
+				[this]()
+				{
+					DynamicReplicatedActor->Destroy();
+					CrossServerSetDynamicReplicatedActor(nullptr);
+					FinishStep();
+				});
+	}
+
+	// Non-replicated Dynamic Actor. Server 1 should have Authority.
+	{
+		AddStep(TEXT("Non-replicated Dynamic Actor - Spawn"), FWorkerDefinition::Server(1), nullptr,
+			[this]()
+			{
+				// Spawning directly on Server 2, but since it's non-replicated it shouldn't migrate to Server 2.
+				DynamicNonReplicatedActor = GetWorld()->SpawnActor<ASpatialAuthorityTestActor>(Server2Position, FRotator::ZeroRotator);
+				FinishStep();
+			});
+
+		AddStep(TEXT("Non-replicated Dynamic Actor - Verify Authority on Server 1"), FWorkerDefinition::Server(1), nullptr, nullptr,
+				[this](float DeltaTime)
+				{
+					if (DynamicNonReplicatedActor->AuthorityOnBeginPlay == 1
+						&& DynamicNonReplicatedActor->AuthorityOnTick == 1)
+					{
+						FinishStep();
+					}
+				});
+
+		float Timer = 0.5f;
+
+		AddStep(TEXT("Non-replicated Dynamic Actor - Verify Dynamic Actor doesn't exist on others"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+				[this, &Timer](float DeltaTime) {
```

Comment:
I might be very wrong about this, but Isn't this potentially causing undefined behaviour, because you capture the reference to a local variable?

Reply:
I was looking at this online and I think you may be right. Not a fan of this :/ I'll change it in a bit to have a class variable.